atomix/cluster/src/main/java/io/atomix/cluster/impl/DiscoveryMembershipProtocol.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.atomix.cluster.impl;
+
+import io.atomix.cluster.BootstrapService;
+import io.atomix.cluster.Member;
+import io.atomix.cluster.MemberId;
+import io.atomix.cluster.discovery.NodeDiscoveryEvent;
+import io.atomix.cluster.discovery.NodeDiscoveryEventListener;
+import io.atomix.cluster.discovery.NodeDiscoveryService;
+import io.atomix.cluster.protocol.GroupMembershipEvent;
+import io.atomix.cluster.protocol.GroupMembershipEventListener;
+import io.atomix.cluster.protocol.GroupMembershipProtocol;
+import io.atomix.cluster.protocol.GroupMembershipProtocolConfig;
+import io.atomix.utils.event.AbstractListenerManager;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link GroupMembershipProtocol} implementation which uses the node discovery service to add and
+ * remove new members. Members can be either initially provided as is, or will be added/removed as
+ * nodes are added/removed in the associated {@link NodeDiscoveryService} passed during {@link
+ * #join(BootstrapService, NodeDiscoveryService, Member)}.
+ */
+public final class DiscoveryMembershipProtocol
```

Comment:
It can be moved, but then cannot be used in tests in Atomix itself, which is always the problem if you have utilities used in both the module and outside. I think it's probably OK to keep as is, so long as it doesn't become complex.